src/nupic/algorithms/TemporalMemory.hpp

```diff
@@ -326,6 +326,15 @@ using namespace nupic::algorithms::connections;
   vector<CellIdx> getWinnerCells() const; //TODO remove?
   void getWinnerCells(sdr::SDR &winnerCells) const;
 
+  /**
+   *  Return TM's output transformed from cells back to columns.
+   *  As a union (logical OR) of active & predicted cells at the
+   *  current time T. 
+   *
+   *  @return SDR size(COLS) 
+   */
+  sdr::SDR getOutputColumns() const;
```

Comment:
better name?

Reply:
I dont think that this is right.  I dont think the TM should ever use the union of active & predicted cells.

Reply:
> There are two use cases for converting from cells to minicolumns, decode and anomaly. We've decided not to implement decode. I would really like to see anomaly as part of this PR. 

Which of the 2 methods do you objectify to? 
- `cellsToColumns`:
  - yes, a convenience method, `columnForCell(cell)` already exists
    - I propose removal of columnForCell
    - or either of them can be made private
  - this method is needed for `TM::anomaly()` which operates with columns.
  - we had the same functionality in VectorHelpers::cellsToColumns
    - but I decided to move it to TM, as it's the only user and the way of conversion (plain topology) makes sense only for TM. So as not to make it confusing and tempting for other, misused usages (if in VectorHelpers) 

- `getOutputColumns` ..TM output as union of active + predictive columns
  - conceptually, I think TM should output only columns (not cells)
    - as cols are needed for it to be used in a hierarchy (and in anomaly, decode,...) 
      - what is Classifier using? It should probably use cols as well.
    - cells should be an implementation detail and hidden from the user
      - cells are still accessible for the likes of CP, etc. 
  - I think we should provide all 3 of them: 
    - getColumnsActive(), Predictive(), Both() 
       - (both = current getOutputCols) 
    - or fill them in compute method `TM::compute(const SDR& input, bool learn, SDR& output)`
      - impl details how to well decide which mode to fill in output (enum?), default param, etc... 
    - Anomaly, NetworkAPI (TMRegion), Classifier(?) would work with these outputs

Reply:
I object to `getOutputColumns`.  It should not mix the active & predictive cells together.

Reply:
>  It should not mix the active & predictive cells together.

why should not? A TM's minicolumn is activated by at least one active cell, be it feed forward activation, or predictive/contextual. 

It's not like it's impossible to compute manually, we have access to the active, predictive cells; and will now have the cellsToColumns transformation. 
  - but the user (for hierarchy, classifier, TMRegion, TM's direct output) has to do exactly what is this method about.

Reply:
Predictive input does not activate mini-columns.  Feed forward input activates mini-columns.

Predictive input can activate a cell within an active mini-column.  If no cells in an active mini-column are predicted then the mini-column bursts and all cells in the mini-column activate.

Reply:
ok, I quickly double checked with https://numenta.com/assets/pdf/biological-and-machine-intelligence/BaMI-Temporal-Memory.pdf 
And you're right, there's no sense in mixing active+pred columns. 

- I'll remove getOutpuColumns
- TMRegion outputs that with option orOutputCols, I then say we should rm that too. 
- instead, do we want to offer TM.activeCols, TM.predictiveCols ? That would be of use for anomaly

Reply:
@ctrl-z-9000-times ok, we agreed that mixing active + predictive cols/cells does not make sense. I'm removing the method from TM, not looking at TMRegion. 
@dkeeney is it necessary for a Region to implement "bottomUpOutput"?

Currently we return either a union of cells (act+pred) or as columns, depending on a switch. 
- what should we return as bottom up output? (I'd say feed-forward activation as columns)
  - or should we remove the field altogether? (ideally) 
- currently, NetworkAPI has now way to obtain columns (active, predicted; separately), and it probably should.

Reply:
> @dkeeney is it necessary for a Region to implement "bottomUpOutput"?

This is its only output.  I think we need something.   I think we could go with columns only.  We could add additional outputs for active and predicted as separate outputs.  
What makes since?

Reply:
any combination of {active, predictive} x {cells, columns} makes sense. 
I'll probably open a new PR for that, and just remove the current orOutputColumns logic

Reply:
See #412  #413